File: src/caco/tui/widgets/library_pane.py
"""Reusable library pane widget with table, info panel, and filtering."""

from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Horizontal, Vertical
from textual.message import Message
from textual.widget import Widget
from textual.widgets import Static

from caco import db
from caco.config import get_tui_config
from caco.db import STATUS_SHORTCUTS
from caco.player import play
from caco.tui.widgets.filter_input import FilterInput
from caco.tui.widgets.sort_select import SortSelect
from caco.tui.widgets.wad_info import WadInfoPanel
from caco.tui.widgets.wad_table import WadTable

# Sort field display names (for notifications)
SORT_DISPLAY = {
    "id": "ID",
    "title": "Title",
    "author": "Author",
    "playtime": "Playtime",
    "last_played": "Last Played",
    "year": "Year",
    "rating": "Rating",
}


class LibraryPane(Widget):
    """Reusable library pane with WAD table, info panel, filtering and sorting.

    This can be used standalone or embedded in tabs with a status filter.
    """

    # Note: j/k/g/G/ctrl+d/ctrl+u are handled by WadTable directly
    BINDINGS = [
        # Filter
        Binding("slash", "focus_filter", "Filter", show=True, key_display="/"),
        Binding("f", "focus_filter", "Filter", show=False),
        # Actions
        Binding("i", "show_details", "Info", show=True),
        Binding("h", "show_history", "History", show=True),
        Binding("e", "edit_wad", "Edit", show=True),
        Binding("d", "delete_wad", "Delete", show=True),
        # Status shortcuts
        Binding("s", "status_mode", "Status", show=True),
        # Sorting
        Binding("o", "focus_sort", "Sort", show=True),
        Binding("O", "toggle_sort_dir", "Sort Dir", show=False, key_display="O"),
        # Rating
        Binding("r", "cycle_rating", "Rating", show=True),
        Binding("R", "clear_rating", "Clear Rating", show=False, key_display="R"),
        # Beaten tracking
        Binding("+", "increment_beaten", "Beaten+", show=False),
        Binding("-", "decrement_beaten", "Beaten-", show=False),
        # Trash view (All tab only)
        Binding("T", "toggle_trash", "Trash", show=False, key_display="T"),
        Binding("u", "restore_wad", "Restore", show=False),
        # Stats
        Binding("S", "show_stats", "Stats", show=False, key_display="S"),
        # Cache
        Binding("C", "show_cache", "Cache", show=False, key_display="C"),
        # Panel toggle
        Binding("P", "toggle_panel", "Panel", show=False, key_display="P"),
        # Escape
        Binding("escape", "escape", "Cancel", show=False),
    ]

    class WadAction(Message):
        """Generic message for actions that need parent handling."""

        def __init__(self, wad_id: int, action: str) -> None:
            super().__init__()
            self.wad_id = wad_id
            self.action = action

    class WadImported(Message):
        """Fired when a WAD is imported (used by parent to refresh other panes)."""

        def __init__(self, wad_id: int) -> None:
            super().__init__()
            self.wad_id = wad_id

    DEFAULT_CSS = """
    LibraryPane {
        height: 100%;
        width: 100%;
    }

    LibraryPane #pane-header {
        height: 3;
        width: 100%;
        padding: 0 1;
        align: left middle;
    }

    LibraryPane #pane-content {
        height: 1fr;
    }

    LibraryPane #wad-list-container {
        width: 2fr;
        height: 100%;
    }

    LibraryPane #info-panel-container {
        width: 1fr;
        height: 100%;
        border-left: solid $primary;
    }

    LibraryPane #status-mode {
        dock: bottom;
        height: 1;
        background: $warning;
        color: $text;
        padding: 0 1;
        display: none;
    }

    LibraryPane #status-mode.visible {
        display: block;
    }
    """

    def __init__(
        self,
        status_filter: str | list[str] | None = None,
        **kwargs,
    ) -> None:
        """Initialize the library pane.

        Args:
            status_filter: If set, pre-filter to only show WADs with this status.
                          Can be a single status string or a list of statuses (OR query).
        """
        super().__init__(**kwargs)
        self._status_filter = status_filter
        self._current_query: str = ""
        self._status_mode = False
        self._show_deleted = False
        tui_config = get_tui_config()
        self._sort_by: str = tui_config.get("default_sort", "id")
        self._sort_desc: bool = tui_config.get("default_sort_desc", False)

    def compose(self) -> ComposeResult:
        with Horizontal(id="pane-header"):
            yield FilterInput(id="filter")
            yield SortSelect(sort_by=self._sort_by, sort_desc=self._sort_desc, id="sort-select")
        with Horizontal(id="pane-content"):
            with Vertical(id="wad-list-container"):
                yield WadTable(id="wad-table")
            with Vertical(id="info-panel-container"):
                yield WadInfoPanel(id="info-panel")
        yield Static("", id="status-mode")

    def on_mount(self) -> None:
        """Load initial data."""
        self._load_wads()
        # Only focus if this is the "All" tab (no status filter)
        # This prevents the last-mounted tab from stealing focus
        if self._status_filter is None:
            self.query_one("#wad-table", WadTable).focus()

    def _get_effective_query(self) -> str | None:
        """Get the effective query combining user query with status filter."""
        parts = []
        if self._status_filter:
            if isinstance(self._status_filter, list):
                # OR query for multiple statuses: "status:a , status:b"
                status_parts = " , ".join(f"status:{s}" for s in self._status_filter)
                parts.append(status_parts)
            else:
                parts.append(f"status:{self._status_filter}")
        if self._current_query:
            parts.append(self._current_query)

        if parts:
            return " ".join(parts)
        return None

    def _load_wads(self) -> None:
        """Load WADs with current query and sort settings."""
        table = self.query_one("#wad-table", WadTable)
        count = table.load_wads(
            self._get_effective_query(),
            sort_by=self._sort_by,
            sort_desc=self._sort_desc,
            include_deleted=self._show_deleted,
        )
        # Update filter bar count
        filter_widget = self.query_one("#filter", FilterInput)
        if self._show_deleted:
            filter_widget.set_wad_count(count, label="trash")
        else:
            filter_widget.set_wad_count(count)

    def refresh_data(self) -> None:
        """Public method to refresh WAD list (called by parent after import)."""
        self._load_wads()

    # -------------------------------------------------------------------------
    # Filter Actions
    # -------------------------------------------------------------------------

    def action_focus_filter(self) -> None:
        """Focus the filter input."""
        filter_widget = self.query_one("#filter", FilterInput)
        filter_widget.show()

    def on_filter_input_query_changed(self, event: FilterInput.QueryChanged) -> None:
        """Handle query change from filter."""
        self._current_query = event.query
        self._load_wads()

    def on_filter_input_escaped(self, event: FilterInput.Escaped) -> None:
        """Handle escape from filter - refocus the table."""
        self.query_one("#wad-table", WadTable).focus()

    def on_filter_input_submitted(self, event: FilterInput.Submitted) -> None:
        """Handle Enter on filter input - refocus the table."""
        self.query_one("#wad-table", WadTable).focus()

    # -------------------------------------------------------------------------
    # Sort Actions
    # -------------------------------------------------------------------------

    def action_focus_sort(self) -> None:
        """Focus the sort dropdown."""
        sort_select = self.query_one("#sort-select", SortSelect)
        select = sort_select.query_one("Select")
        select.focus()

    def on_sort_select_sort_changed(self, event: SortSelect.SortChanged) -> None:
        """Handle sort change from dropdown."""
        self._sort_by = event.sort_by
        self._sort_desc = event.sort_desc
        self._load_wads()
        direction = "↓" if self._sort_desc else "↑"
        display_name = SORT_DISPLAY.get(self._sort_by, self._sort_by)
        self.notify(f"Sort: {display_name} {direction}")
        # Return focus to table after sort change
        self.query_one("#wad-table", WadTable).focus()

    def action_toggle_sort_dir(self) -> None:
        """Toggle sort direction (keyboard shortcut)."""
        sort_select = self.query_one("#sort-select", SortSelect)
        sort_select.toggle_direction()

    # -------------------------------------------------------------------------
    # WAD Selection/Activation Events
    # -------------------------------------------------------------------------

    def on_wad_table_wad_selected(self, event: WadTable.WadSelected) -> None:
        """Update info panel when WAD selection changes."""
        panel = self.query_one("#info-panel", WadInfoPanel)
        if event.wad_id is not None:
            table = self.query_one("#wad-table", WadTable)
            stats = table.get_wad_stats(event.wad_id)
            panel.update_wad(event.wad_id, stats=stats)
        else:
            panel.update_wad(None)

    def on_wad_table_wad_activated(self, event: WadTable.WadActivated) -> None:
        """Handle Enter on a WAD row - play the WAD."""
        if self._show_deleted:
            self.notify("Cannot play a deleted WAD", severity="warning")
            return
        self._play_wad(event.wad_id)

    # -------------------------------------------------------------------------
    # Play Action
    # -------------------------------------------------------------------------

    def _play_wad(self, wad_id: int) -> None:
        """Play a WAD by ID."""
        wad = db.get_wad(wad_id)
        if not wad:
            self.notify("WAD not found", severity="error")
            return

        self.notify(f"Launching {wad['title']}...")
        self.run_worker(self._play_and_refresh(wad_id))

    async def _play_and_refresh(self, wad_id: int) -> None:
        """Play WAD and refresh after."""
        wad = db.get_wad(wad_id)
        if not wad:
            return

        error = None
        with self.app.suspend():
            try:
                play(wad_id)
            except ValueError as e:
                error = str(e)

        if error:
            self.notify(error, severity="error", timeout=10)
        else:
            self.notify(f"Finished playing {wad['title']}")

        self._load_wads()
        panel = self.query_one("#info-panel", WadInfoPanel)
        panel.update_wad(wad_id)

    # -------------------------------------------------------------------------
    # Detail/Edit Screens
    # -------------------------------------------------------------------------

    def action_show_details(self) -> None:
        """Show WAD detail screen."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        from caco.tui.screens.wad_detail import WadDetailScreen

        self.app.push_screen(WadDetailScreen(wad_id))

    def action_show_history(self) -> None:
        """Show session history screen."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        from caco.tui.screens.sessions import SessionsScreen

        self.app.push_screen(SessionsScreen(wad_id))

    def action_edit_wad(self) -> None:
        """Open WAD edit screen."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        from caco.tui.screens.wad_edit import WadEditScreen

        def on_dismiss(result: bool | None) -> None:
            if result:
                self._load_wads()
                panel = self.query_one("#info-panel", WadInfoPanel)
                panel.update_wad(wad_id)

        self.app.push_screen(WadEditScreen(wad_id), on_dismiss)

    # -------------------------------------------------------------------------
    # Delete
    # -------------------------------------------------------------------------

    def action_delete_wad(self) -> None:
        """Delete the selected WAD (with confirmation)."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        from caco.tui.screens.confirm_delete import ConfirmDeleteScreen

        def on_dismiss(result: int | None) -> None:
            if result is not None:
                db.delete_wad(result)
                self.notify("Moved to trash")
                self._load_wads()
                self.post_message(self.WadImported(result))

        self.app.push_screen(ConfirmDeleteScreen(wad_id), on_dismiss)

    # -------------------------------------------------------------------------
    # Rating
    # -------------------------------------------------------------------------

    def action_cycle_rating(self) -> None:
        """Cycle rating 0→1→2→3→4→5→0 for selected WAD."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        wad = db.get_wad(wad_id)
        if not wad:
            return

        current = wad.get("rating") or 0
        new_rating = (current + 1) % 6
        if new_rating == 0:
            db.update_wad(wad_id, rating=None)
            self.notify("Rating cleared")
        else:
            db.update_wad(wad_id, rating=new_rating)
            stars = "★" * new_rating + "☆" * (5 - new_rating)
            self.notify(f"Rating: {stars}")

        table.update_row(wad_id)
        panel = self.query_one("#info-panel", WadInfoPanel)
        panel.update_wad(wad_id)

    def action_clear_rating(self) -> None:
        """Clear rating for selected WAD."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        db.update_wad(wad_id, rating=None)
        table.update_row(wad_id)
        panel = self.query_one("#info-panel", WadInfoPanel)
        panel.update_wad(wad_id)
        self.notify("Rating cleared")

    # -------------------------------------------------------------------------
    # Beaten Tracking
    # -------------------------------------------------------------------------

    def action_increment_beaten(self) -> None:
        """Increment beaten count for selected WAD."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        db.add_wad_completion(wad_id)
        count = db.get_times_beaten(wad_id)
        table.update_row(wad_id)
        panel = self.query_one("#info-panel", WadInfoPanel)
        panel.update_wad(wad_id)
        self.notify(f"Beaten: {count}x")

    def action_decrement_beaten(self) -> None:
        """Decrement beaten count for selected WAD."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        completions = db.get_wad_completions(wad_id)
        if not completions:
            self.notify("No completions to remove", severity="warning")
            return

        # Delete most recent completion
        db.delete_wad_completion(completions[0]["id"])
        count = db.get_times_beaten(wad_id)
        table.update_row(wad_id)
        panel = self.query_one("#info-panel", WadInfoPanel)
        panel.update_wad(wad_id)
        self.notify(f"Beaten: {count}x")

    # -------------------------------------------------------------------------
    # Status Mode
    # -------------------------------------------------------------------------

    def action_status_mode(self) -> None:
        """Enter status mode for quick status change."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        self._status_mode = True
        status_indicator = self.query_one("#status-mode", Static)
        status_indicator.update(
            "Status: [p]laying [f]inished [t]o-play [b]acklog [a]bandoned [w]aiting"
        )
        status_indicator.add_class("visible")

    def on_key(self, event) -> None:
        """Handle key events for status mode."""
        if self._status_mode:
            key = event.key
            if key in STATUS_SHORTCUTS:
                self._set_status(STATUS_SHORTCUTS[key])
                event.stop()
            elif key == "escape":
                self._exit_status_mode()
                event.stop()
            else:
                self._exit_status_mode()

    def _set_status(self, status: str) -> None:
        """Set status of selected WAD."""
        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id:
            db.update_wad(wad_id, status=status)
            self.notify(f"Status set to {status}")

            # If we have a status filter and the WAD no longer matches, do full reload
            if self._status_filter:
                filters = self._status_filter if isinstance(self._status_filter, list) else [self._status_filter]
                if status not in filters:
                    self._load_wads()
                else:
                    table.update_row(wad_id)
            else:
                table.update_row(wad_id)

            panel = self.query_one("#info-panel", WadInfoPanel)
            panel.update_wad(wad_id)
            # Notify parent that WAD data changed (for other tab refresh)
            self.post_message(self.WadImported(wad_id))

        self._exit_status_mode()

    def _exit_status_mode(self) -> None:
        """Exit status mode."""
        self._status_mode = False
        status_indicator = self.query_one("#status-mode", Static)
        status_indicator.remove_class("visible")
        status_indicator.update("")

    # -------------------------------------------------------------------------
    # Trash View
    # -------------------------------------------------------------------------

    def action_toggle_trash(self) -> None:
        """Toggle trash/deleted view (All tab only)."""
        if self._status_filter is not None:
            self.notify("Trash view is only available in the All tab", severity="warning")
            return

        self._show_deleted = not self._show_deleted
        self._load_wads()

        if self._show_deleted:
            self.notify("Showing trash (u=Restore)")
        else:
            self.notify("Showing library")

    def action_restore_wad(self) -> None:
        """Restore a deleted WAD from trash."""
        if not self._show_deleted:
            return

        table = self.query_one("#wad-table", WadTable)
        wad_id = table.get_selected_wad_id()

        if wad_id is None:
            self.notify("No WAD selected", severity="warning")
            return

        if db.restore_wad(wad_id):
            self.notify("WAD restored from trash")
            self._load_wads()
            self.post_message(self.WadImported(wad_id))
        else:
            self.notify("Failed to restore WAD", severity="error")

    # -------------------------------------------------------------------------
    # Stats Screen
    # -------------------------------------------------------------------------

    def action_show_stats(self) -> None:
        """Show library stats screen."""
        from caco.tui.screens.stats import StatsScreen
        self.app.push_screen(StatsScreen())

    # -------------------------------------------------------------------------
    # Cache Screen
    # -------------------------------------------------------------------------

    def action_show_cache(self) -> None:
        """Show cache management screen."""
        from caco.tui.screens.cache import CacheScreen
        self.app.push_screen(CacheScreen())

    # -------------------------------------------------------------------------
    # Panel Toggle & Responsive Layout
    # -------------------------------------------------------------------------

    def action_toggle_panel(self) -> None:
        """Toggle info panel visibility."""
        container = self.query_one("#info-panel-container", Vertical)
        container.display = not container.display

    def on_resize(self, event) -> None:
        """Auto-hide info panel on narrow terminals."""
        container = self.query_one("#info-panel-container", Vertical)
        if event.size.width < 100:
            container.display = False
        else:
            container.display = True

    # -------------------------------------------------------------------------
    # Escape
    # -------------------------------------------------------------------------

    def action_escape(self) -> None:
        """Handle escape key."""
        if self._status_mode:
            self._exit_status_mode()
        else:
            filter_widget = self.query_one("#filter", FilterInput)
            if filter_widget.display:
                filter_widget.clear()
                self.query_one("#wad-table", WadTable).focus()

File: crates/caco-tui/src/widgets/library_pane.rs
use std::time::Instant;

use caco_core::db;
use caco_core::db::models::Status;
use caco_core::db::sessions;
use caco_core::db::wads;
use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::Frame;
use rusqlite::Connection;

use crate::message::{AppMessage, ScreenId, Severity};
use crate::widgets::filter_input::{self, FilterInputState};
use crate::widgets::sort_select::{self, SortSelectState};
use crate::widgets::status_bar;
use crate::widgets::wad_info::{self, WadInfoState};
use crate::widgets::wad_table::{self, WadTableState};

const GG_TIMEOUT_MS: u128 = 500;

/// State for the library pane — composes all sub-widget states.
pub struct LibraryPaneState {
    pub table: WadTableState,
    pub filter: FilterInputState,
    pub sort: SortSelectState,
    pub info: WadInfoState,
    /// Tab query filter (e.g. "intent:inbox", "play:started"), or None for "All".
    pub tab_query: Option<String>,
    pub status_mode: bool,
    pub show_panel: bool,
    pub show_trash: bool,
}

impl LibraryPaneState {
    pub fn new(
        tab_query: Option<String>,
        sort_field: &str,
        sort_desc: bool,
    ) -> Self {
        Self {
            table: WadTableState::new(),
            filter: FilterInputState::new(),
            sort: SortSelectState::from_config(sort_field, sort_desc),
            info: WadInfoState::new(),
            tab_query,
            status_mode: false,
            show_panel: true,
            show_trash: false,
        }
    }

    /// Build the effective query string combining tab query filter and user query.
    fn effective_query(&self) -> Option<String> {
        let mut parts = Vec::new();

        if let Some(ref tab_q) = self.tab_query {
            parts.push(tab_q.clone());
        }

        let user_query = self.filter.query();
        if !user_query.is_empty() {
            parts.push(user_query.to_string());
        }

        if parts.is_empty() {
            None
        } else {
            Some(parts.join(" "))
        }
    }

    /// Reload WADs from the database.
    pub fn reload(&mut self, conn: &Connection) {
        let query = self.effective_query();
        let count = self.table.load_wads(
            conn,
            query.as_deref(),
            Some(self.sort.field()),
            self.sort.sort_desc,
            self.show_trash,
        );
        self.filter.wad_count = count;
    }

    /// Handle key events. Returns `Some(AppMessage)` for actions that need app-level handling.
    pub fn handle_key(&mut self, key: KeyEvent, conn: &Connection) -> Option<AppMessage> {
        // Status mode captures single key
        if self.status_mode {
            return self.handle_status_mode_key(key, conn);
        }

        // Filter captures all keys when focused
        if self.filter.focused {
            if let Some(_query) = self.filter.handle_key(key) {
                self.reload(conn);
            }
            return None;
        }

        match (key.code, key.modifiers) {
            // Filter
            (KeyCode::Char('/') | KeyCode::Char('f'), KeyModifiers::NONE) => {
                self.filter.focused = true;
                None
            }

            // Vim navigation
            (KeyCode::Char('j') | KeyCode::Down, KeyModifiers::NONE) => {
                self.table.next();
                None
            }
            (KeyCode::Char('k') | KeyCode::Up, KeyModifiers::NONE) => {
                self.table.previous();
                None
            }
            (KeyCode::Char('g'), KeyModifiers::NONE) => {
                // gg = go to top
                if let Some(first_g) = self.table.g_pressed {
                    if first_g.elapsed().as_millis() < GG_TIMEOUT_MS {
                        self.table.first();
                        self.table.g_pressed = None;
                    } else {
                        self.table.g_pressed = Some(Instant::now());
                    }
                } else {
                    self.table.g_pressed = Some(Instant::now());
                }
                None
            }
            (KeyCode::Char('G'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                self.table.last();
                None
            }
            (KeyCode::Char('d'), KeyModifiers::CONTROL) => {
                self.table.page_down(20);
                None
            }
            (KeyCode::Char('u'), KeyModifiers::CONTROL) => {
                self.table.page_up(20);
                None
            }

            // Actions
            (KeyCode::Char('i'), KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PushScreen(ScreenId::WadDetail(id)))
                } else {
                    None
                }
            }
            (KeyCode::Char('h'), KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PushScreen(ScreenId::Sessions(id)))
                } else {
                    None
                }
            }
            (KeyCode::Char('e'), KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PushScreen(ScreenId::WadEdit(id)))
                } else {
                    None
                }
            }
            (KeyCode::Char('d'), KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PushScreen(ScreenId::ConfirmDelete(id)))
                } else {
                    None
                }
            }

            // Status mode
            (KeyCode::Char('s'), KeyModifiers::NONE) => {
                if self.table.selected_wad_id().is_some() {
                    self.status_mode = true;
                }
                None
            }

            // Sort
            (KeyCode::Char('o'), KeyModifiers::NONE) => {
                self.sort.cycle();
                self.reload(conn);
                None
            }
            (KeyCode::Char('O'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                self.sort.toggle_direction();
                self.reload(conn);
                None
            }

            // Rating
            (KeyCode::Char('r'), KeyModifiers::NONE) => {
                self.cycle_rating(conn);
                None
            }
            (KeyCode::Char('R'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                self.clear_rating(conn);
                None
            }

            // Beaten
            (KeyCode::Char('+') | KeyCode::Char('='), _) => {
                self.add_beaten(conn);
                None
            }
            (KeyCode::Char('-'), KeyModifiers::NONE) => {
                self.remove_beaten(conn);
                None
            }

            // Map stats
            (KeyCode::Char('M'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PushScreen(ScreenId::WadStats(id)))
                } else {
                    None
                }
            }

            // Trash
            (KeyCode::Char('T'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                self.show_trash = !self.show_trash;
                self.reload(conn);
                None
            }
            (KeyCode::Char('u'), KeyModifiers::NONE) => {
                if self.show_trash {
                    if let Some(id) = self.table.selected_wad_id() {
                        let _ = wads::restore_wad(conn, id);
                        self.reload(conn);
                        return Some(AppMessage::Notify(
                            format!("WAD #{id} restored"),
                            Severity::Info,
                        ));
                    }
                }
                None
            }

            // Screens
            (KeyCode::Char('S'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                Some(AppMessage::PushScreen(ScreenId::Stats))
            }
            (KeyCode::Char('C'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                Some(AppMessage::PushScreen(ScreenId::Cache))
            }
            (KeyCode::Char('W'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                Some(AppMessage::PushScreen(ScreenId::Resources))
            }

            // Panel toggle
            (KeyCode::Char('P'), KeyModifiers::SHIFT | KeyModifiers::NONE) => {
                self.show_panel = !self.show_panel;
                None
            }

            // Play
            (KeyCode::Enter, KeyModifiers::NONE) => {
                if let Some(id) = self.table.selected_wad_id() {
                    Some(AppMessage::PlayWad(id))
                } else {
                    None
                }
            }

            _ => None,
        }
    }

    fn handle_status_mode_key(
        &mut self,
        key: KeyEvent,
        conn: &Connection,
    ) -> Option<AppMessage> {
        self.status_mode = false;

        let status = match key.code {
            KeyCode::Char('p') => "playing",
            KeyCode::Char('f') => "finished",
            KeyCode::Char('t') => "to-play",
            KeyCode::Char('b') => "backlog",
            KeyCode::Char('a') => "abandoned",
            KeyCode::Char('w') => "awaiting-update",
            KeyCode::Esc => return None,
            _ => return None,
        };

        if let Some(id) = self.table.selected_wad_id() {
            if let Ok(update) = db::wads::WadUpdate::new()
                .set_status(Status::parse(status).unwrap_or(Status::ToPlay))
            {
                let _ = wads::update_wad(conn, id, &update);
                self.table.update_row(conn, id);
                return Some(AppMessage::Notify(
                    format!("Status → {}", crate::theme::status_display(status)),
                    Severity::Info,
                ));
            }
        }
        None
    }

    fn cycle_rating(&mut self, conn: &Connection) {
        if let Some(wad) = self.table.selected_wad() {
            let current = wad.rating.unwrap_or(0);
            let new_rating = if current >= 5 { 0 } else { current + 1 };
            let value = if new_rating == 0 {
                None
            } else {
                Some(new_rating as i64)
            };
            if let Ok(update) = db::wads::WadUpdate::new().set_int("rating", value) {
                let _ = wads::update_wad(conn, wad.id, &update);
                self.table.update_row(conn, wad.id);
            }
        }
    }

    fn clear_rating(&mut self, conn: &Connection) {
        if let Some(wad) = self.table.selected_wad() {
            if let Ok(update) = db::wads::WadUpdate::new().set_int("rating", None) {
                let _ = wads::update_wad(conn, wad.id, &update);
                self.table.update_row(conn, wad.id);
            }
        }
    }

    fn add_beaten(&mut self, conn: &Connection) {
        if let Some(id) = self.table.selected_wad_id() {
            let _ = sessions::add_wad_completion(conn, id, None, None, None);
            self.table.update_row(conn, id);
        }
    }

    fn remove_beaten(&mut self, conn: &Connection) {
        if let Some(id) = self.table.selected_wad_id() {
            if let Ok(completions) = sessions::get_wad_completions(conn, id) {
                if let Some(last) = completions.last() {
                    let _ = sessions::delete_wad_completion(conn, last.id);
                    self.table.update_row(conn, id);
                }
            }
        }
    }

    /// Called on tick to check filter debounce.
    pub fn tick(&mut self, conn: &Connection) -> Option<AppMessage> {
        if let Some(_query) = self.filter.tick() {
            self.reload(conn);
        }
        None
    }
}

/// Render the library pane.
pub fn render_library_pane(
    state: &mut LibraryPaneState,
    frame: &mut Frame,
    area: Rect,
    terminal_width: u16,
) {
    // Layout: header (filter+sort) | content | status bar
    let show_status_mode = state.status_mode;
    let bottom_height = if show_status_mode { 1 } else { 1 };

    let layout = Layout::vertical([
        Constraint::Length(1), // filter + sort
        Constraint::Min(1),   // content
        Constraint::Length(bottom_height),  // status/hints
    ])
    .split(area);

    let header_area = layout[0];
    let content_area = layout[1];
    let bottom_area = layout[2];

    // Header: filter (left) + sort (right)
    let header_layout = Layout::horizontal([
        Constraint::Min(20),
        Constraint::Length(25),
    ])
    .split(header_area);

    filter_input::render_filter_input(&state.filter, frame, header_layout[0]);
    sort_select::render_sort_select(&state.sort, frame, header_layout[1]);

    // Content: table (left) + info panel (right)
    let show_panel = state.show_panel && terminal_width >= 100;
    if show_panel {
        let content_layout = Layout::horizontal([
            Constraint::Percentage(65),
            Constraint::Percentage(35),
        ])
        .split(content_area);

        wad_table::render_wad_table(&mut state.table, frame, content_layout[0]);

        let wad = state.table.selected_wad();
        let stats = state.table.selected_stats();
        wad_info::render_wad_info(&mut state.info, wad, stats, frame, content_layout[1]);
    } else {
        wad_table::render_wad_table(&mut state.table, frame, content_area);
    }

    // Bottom: status mode bar or key hints
    if show_status_mode {
        status_bar::render_status_mode_bar(frame, bottom_area);
    } else {
        let hints = if state.show_trash {
            vec![
                ("j/k", "nav"),
                ("u", "restore"),
                ("T", "exit trash"),
                ("q", "quit"),
            ]
        } else {
            vec![
                ("j/k", "nav"),
                ("/", "filter"),
                ("i", "info"),
                ("e", "edit"),
                ("s", "status"),
                ("Enter", "play"),
                ("q", "quit"),
            ]
        };
        status_bar::render_status_bar(&hints, frame, bottom_area);
    }
}
